### PR TITLE
emacs 27.1

### PR DESCRIPTION
**Pull request by CamJN**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

there were issues with the audit, listed below, not sure how to fix them:

```
emacs:
  * Formula emacs contains deprecated SPDX licenses: ["GPL-3.0"].
  * Files were found with references to the Homebrew shims directory.
    The offending files are:
      libexec/emacs/27.1/x86_64-apple-darwin19.6.0/emacs.pdmp
```

### CODE REVIEW

**Comment by miccal** [comment]
I can answer one of your questions:
```
* Formula emacs contains deprecated SPDX licenses: ["GPL-3.0"].
```
The `license` should be:
```
license "GPL-3.0-or-later"
```

**Comment by aconchillo** [comment]
Seems to be affected by https://github.com/Homebrew/homebrew-core/issues/58984.

**Comment by CamJN** [comment]
@aconchillo can you elaborate how so? I looked at the issue you linked and it looks like a Haskell sandbox issue?

**Comment by aconchillo** [comment]
Yes, there's been other packages unrelated to Haskell AFAICT with the same issue. For example: https://github.com/Homebrew/homebrew-core/pull/58683. I don't know how `agda` has nothing to do with all this, but I'm experiencing the same issue in this PR https://github.com/Homebrew/homebrew-core/pull/59490.

**Comment by gromgit** [comment]
Version-bumping any formula requires testing all the formulae that depend on it:
```
$ brew uses --include-build --recursive emacs
agda                   cask                   cedille                doxymacs               eless                  emacs-dracula          gforth                 lfe                    mu                     nesc                   proof-general          rtags
```

**Comment by aconchillo** [comment]
Thanks @gromgit. That makes sense and I guess it explains the long times of the check.

**Comment by aconchillo** [comment]
This PR https://github.com/Homebrew/homebrew-core/pull/59709 will fix the issue once merged.

**Comment by chenrui333** [comment]
@CamJN can you rebased your PR, I just merged the Agda fix PR.

**Comment by CamJN** [comment]
Still getting this:

```
=> FAILED
40
Error: 1 problem in 1 formula detected
41
emacs:
42
  * Files were found with references to the Homebrew shims directory.
43
    The offending files are:
44
      libexec/emacs/27.1/x86_64-apple-darwin19.5.0/emacs.pdmp
45

```

I think this is due to emacs now using a new portable dumper instead of the abomination that was `unexec`.

**Comment by CamJN** [comment]
Indeed, the portable dump file does refer to /usr/local/Homebrew/Library/Homebrew/shims/mac/super according to `strings`.

How does homebrew usually deal with things like this?

**Comment by CamJN** [comment]
Some discussion here: https://archive.casouri.cat/note/2020/painless-transition-to-portable-dumper/index.html

**Comment by CamJN** [comment]
I rebuilt using the `--with-dumping=unexec` configure option, and ran `find . -type f -exec fgrep -ai '/shims' {} \; -print` which seems to not include a reference to the shims dir, except in the `config.log` file.

That said, that's not a viable strategy going forward as `unexec` is going away eventually.

**Comment by SMillerDev** [comment]
What files does it reference?

**Comment by CamJN** [comment]
@SMillerDev it just refers to the path, in fact I believe (but don't know for sure yet) it is just capturing the PATH env var at the time it dumps the new executable.

**Comment by CamJN** [comment]
@SMillerDev here's the strings output that makes me think it's just the `PATH`:

```
$ strings /usr/local/opt/emacs/libexec/emacs/27.1/x86_64-apple-darwin19.6.0/emacs.pdmp | rg -B 13 shims
/sbin
/usr/sbin
/bin
/usr/bin
/usr/local/opt/gnutls/bin
/usr/local/opt/libevent/bin
/usr/local/opt/openssl@1.1/bin
/usr/local/opt/p11-kit/bin
/usr/local/opt/nettle/bin
/usr/local/opt/libtasn1/bin
/usr/local/opt/libidn2/bin
/usr/local/opt/gettext/bin
/usr/local/opt/pkg-config/bin
/usr/local/Homebrew/Library/Homebrew/shims/mac/super
```

**Comment by aconchillo** [comment]
This is not directly related but in https://github.com/Homebrew/homebrew-core/pull/60307#discussion_r478917491 I believe I explained why Homebrew does this type of checks. Basically, when a bottle is built Homebrew doesn't want references to Homebrew prefix hardcoded in files, this is so at installation time a new prefix can be used.

Homebrew takes care of this in text files when building a bottle by replacing Homebrew path references with placeholders  and replaces them back with the Homebrew prefix at installation time, but in this case it's a binary file so it can't do any of that.

I don't really know enough about Homebrew but this seems a tough one. Without knowing much, the only thing that comes to my mind is creating a link to the shims directory:

```
/usr/local/opt/homebrew-shims/ -> /usr/local/Homebrew/Library/Homebrew/shims/mac/super
```

or something like that and add the `/usr/local/opt/...` directory to PATH instead of `/usr/local/Homebrew/...`. This way the formula shims check will succeed and bottles will be able to install in different Homebrew prefixes as long as `/usr/local/opt` is fixed (i.e. the same in system where bottle is built and system where bottle is installed).

Hopefully there's a simpler solution to this. I use emacs in the terminal on my day to day work, so I'm very interested in this one :-D.

**Comment by CamJN** [comment]
I've been busy, but I have a theory that we can edit the PATH by loading some elisp before the executable is dumped. I'm going to try but I can't do it right now.

**Comment by aconchillo** [comment]
> I've been busy, but I have a theory that we can edit the PATH by loading some elisp before the executable is dumped. I'm going to try but I can't do it right now.

Great! Would that remove `/usr/local/Homebrew/Library/Homebrew/shims/mac/super`?

**Comment by CamJN** [comment]
@aconchillo that is the plan, yes.

**Comment by aconchillo** [comment]
I was curious and tried combinations of the following (this is just a test):

```
;; Remove Homebrew shims path from PATH since Homebrew doesn't
;; like having hardcoded paths to the Homebrew directories.
;; See https://github.com/Homebrew/homebrew-core/pull/59448#issue-465829006
(let ((env-path (getenv "PATH")))
  (message "Before PATH is %s" env-path)
  (message "Process environment is %s"
           (mapconcat 'identity initial-environment " "))
  (if env-path
      (setenv "PATH"
              (mapconcat 'identity
                         (delete "/usr/local/Homebrew/Library/Homebrew/shims/mac/super"
                                 (split-string env-path ":"))
                         ":")))
  (message "After PATH is %s" (getenv "PATH")))
```

in different parts of `loadup.el` without success. `loadup.el` is where emacs.pdmp is created, but `(getenv "PATH")` returns nil. I'll investigate more.

**Comment by CamJN** [comment]
Rather than edit loadup.el, you want to add a `site-load.el` to the `lisp/` dir. And yes, the `getenv` is where I last had issues.

**Comment by CamJN** [comment]
Ok, I got rid of the reference to the homebrew shims dir, someone who is up to date on what homebrew wants stylistically should chime in on how they want this file added.

File: Formula/emacs.rb
Comment (SMillerDev) [inline Formula/emacs.rb:107]: Could this be done as something like
```ruby
("list/site-load.el").write <~EOS
+(setq exec-path (delete nil (mapcar (lambda (elt) (unless (string-match-p "Homebrew/shims" elt) elt)) exec-path)))
EOS
```

File: Formula/emacs.rb
Comment (SMillerDev) [inline Formula/emacs.rb:40]: Is this needed? It seems unrelated

Comment (CamJN) [reply Formula/emacs.rb:40]: That configure flag is no longer supported. While I was testing I saw that the configure script warns that it's unrecognized and so I removed it.

File: Formula/emacs.rb
Comment (CamJN) [reply Formula/emacs.rb:107]: Yeah, I didn't really know what the preferred method was here, was just trying to fix this: https://github.com/Homebrew/homebrew-core/runs/1067888651 I'll push a fix.

Comment (CamJN) [reply Formula/emacs.rb:107]: @SMillerDev brew audit doesn't seem to like this:

```
    ("lisp/site-load.el").write <~EOS
    (setq exec-path (delete nil
                     (mapcar
                      (lambda (elt)
                       (unless (string-match-p "Homebrew/shims" elt) elt))
                      exec-path)))
    EOS
```

gives:

```
brew audit --strict  ~/Developer/Ruby/homebrew-core/Formula/emacs.rb
Error: emacs: /Users/camdennarzt/Developer/Ruby/homebrew-core/Formula/emacs.rb:60: syntax error, unexpected ( arg, expecting do or '{' or '('
    (setq exec-path (delete nil
                    ^
/Users/camdennarzt/Developer/Ruby/homebrew-core/Formula/emacs.rb:61: syntax error, unexpected '(', expecting ')'
                     (mapcar
                     ^
/Users/camdennarzt/Developer/Ruby/homebrew-core/Formula/emacs.rb:63: syntax error, unexpected tSTRING_BEG, expecting do or '{' or '('
...      (unless (string-match-p "Homebrew/shims" elt) elt))
...                              ^
/Users/camdennarzt/Developer/Ruby/homebrew-core/Formula/emacs.rb:63: syntax error, unexpected tIDENTIFIER, expecting ')'
...ng-match-p "Homebrew/shims" elt) elt))
...                            ^~~
/Users/camdennarzt/Developer/Ruby/homebrew-core/Formula/emacs.rb:64: syntax error, unexpected ')', expecting end
...                    exec-path)))
...                              ^
```

Comment (CamJN) [reply Formula/emacs.rb]: done

**Comment by aconchillo** [comment]
@CamJN Oh... so it was `exec-path`. Nice!

File: Formula/emacs.rb
Comment (aconchillo) [inline Formula/emacs.rb:59]: this should probably be?

```
(prefix/"lisp/site-load.el").write
```

**Comment by CamJN** [comment]
Sweet, all checks passed. Do I need to squash or can you guys use the squash and merge button?

**Comment by SMillerDev** [comment]
Not without breaking things. Please squash the commits.

**Comment by CamJN** [comment]
Done

**Comment by CamJN** [comment]
`agda` timed out on 10.15 but the formula change is the same that passed everything last time.

**Comment by CamJN** [comment]
@SMillerDev is there a way to run just the failed test again? Or kick off all the tests? Not sure what to do about the `agda` test since that was supposed to be fixed and I rebased on master so I should have that commit included.

**Comment by CamJN** [comment]
@chenrui333 any idea how to speed up the agda tests? They're timing out: https://github.com/Homebrew/homebrew-core/pull/59448/checks?check_run_id=1070317115#step:5:699

**Comment by aconchillo** [comment]
> @SMillerDev is there a way to run just the failed test again? Or kick off all the tests? Not sure what to do about the `agda` test since that was supposed to be fixed and I rebased on master so I should have that commit included.

Rebasing and pushing again should kick off all tests.

**Comment by CamJN** [comment]
All checks passed again.

**Review by SMillerDev — APPROVED** [review]
Thanks @CamJN  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Pull request closed without merging**
